### PR TITLE
i#2626 AArch64 encoder: Add support for indexed vector instrs.

### PR DESCRIPTION
**Pull request by fhahn**
This patch adds the operands required to add indexed vector
instructions to codec.txt and adds the vector-by-element encoding of
FMLA as an example. Additional tests and macros should be added once
the script in the project-aarch64-generate-patterns branch gets updated.

Issue #2626

### CODE REVIEW

File: core/arch/aarch64/codec.c [hunk 1]
```diff
@@ -940,6 +940,33 @@ encode_opnd_dq5(uint enc, int opcode, byte *pc, opnd_t opnd, OUT uint *enc_out)
     return encode_opnd_dq_plus(0, 5, 30, opnd, enc_out);
 }
 
+/*
+ * dq16: D/Q register at bit position 16 with 4 bits only, for the FP16
```
Comment (egrimley) [inline core/arch/aarch64/codec.c]: This should be "dq16_fsz16"?

File: core/arch/aarch64/codec.c [hunk 2]
```diff
@@ -940,6 +940,33 @@ encode_opnd_dq5(uint enc, int opcode, byte *pc, opnd_t opnd, OUT uint *enc_out)
...
+ *       by-element encoding; bit 30 selects Q reg */
```
Comment (egrimley) [inline core/arch/aarch64/codec.c]: `*\` should be on next line.

File: core/arch/aarch64/codec.txt [hunk 3]
```diff
@@ -124,6 +124,11 @@
 ??--------xxxxxxxxxxxxxxxxx-----  mem12      # gets size from 31:30
 ----------xxxxxxxxxxxxxxxxx-----  mem12q     # size is 16 bytes
 ----------xxxxxxxxxxxxxxxxx-----  prf12      # size is 0 bytes (prefetch variant of mem12)
+----------xx--------x-----------  vindex_H   # Index for vector with half elements (0-7)
```
Comment (egrimley) [inline core/arch/aarch64/codec.txt]: Though one may feel that these belong together, they are not correctly ordered by pattern.

File: suite/tests/api/dis-a64.txt [hunk 4]
```diff
@@ -1782,6 +1782,29 @@ fd7fffff : ldr    d31, [sp,#32760]        : ldr    +0x7ff8(%sp)[8byte] -> %d31
 4e7bcc8c : fmla v12.2d, v4.2d, v27.2d : fmla   %q12 %q4 %q27 $0x03 -> %q12
 0e3bcc8c : fmla v12.2s, v4.2s, v27.2s : fmla   %d12 %d4 %d27 $0x02 -> %d12
 
+# FMLA (by element)
```
Comment (egrimley) [inline suite/tests/api/dis-a64.txt]: Very trivial nit: Why the multiple spaces in these lines?

File: core/arch/aarch64/codec.c (hunk 1 above)
Comment (fhahn) [reply core/arch/aarch64/codec.c]: Yep, done. I am not entirely happy with the name fsz16 overall, I think we could have a better name that makes it more clear that it expresses the element width for half precision. But that's for another change.

File: core/arch/aarch64/codec.c [hunk 5]
```diff
@@ -1739,6 +1766,89 @@ encode_opnd_prf12(uint enc, int opcode, byte *pc, opnd_t opnd, OUT uint *enc_out
...
+    return true;
+}
+
+/* vindex_S: Index for vector with single elements (0-3). */
+
+static inline bool
+decode_opnd_vindex_S(uint enc, int opcode, byte *pc, OUT opnd_t *opnd)
+{
+    uint bits;
+    ASSERT((enc >> 22 & 1) == 0 && "vindex_S should only be used for single width.");
```
Comment (egrimley) [inline core/arch/aarch64/codec.c]: I don't understand why this isn't just a "return false". There is no other ASSERT in this set of functions. I wonder if there's a better way of handling this bit 22. Is it possible to combine vindex_S and vindex_D, with a '?' in bit 30 in the pattern in `codec.txt`?

Comment (fhahn) [reply core/arch/aarch64/codec.c]: Right, I've updated it to use a single vindex_SD with '?'. I missed the fact that the codec.py script supports reading bits sets by other operands. Neat!

File: suite/tests/api/dis-a64.txt (hunk 4 above)
Comment (fhahn) [reply suite/tests/api/dis-a64.txt]: Removed

File: core/arch/aarch64/codec.txt (hunk 3 above)
Comment (fhahn) [reply core/arch/aarch64/codec.txt]: Reordered

File: core/arch/aarch64/codec.c [hunk 6]
```diff
@@ -940,6 +940,33 @@ encode_opnd_dq5(uint enc, int opcode, byte *pc, opnd_t opnd, OUT uint *enc_out)
     return encode_opnd_dq_plus(0, 5, 30, opnd, enc_out);
 }
 
+/*
```
Comment (egrimley) [inline core/arch/aarch64/codec.c]: Wrong comment style here! (Is there a script that checks the way comments are formatted? Of course, the licence notices do not follow the prescribed style.)

Is it safe to approve the commit anyway? I don't think it'll get pushed automatically...

Comment (fhahn) [reply core/arch/aarch64/codec.c]: Thanks that should be fixed now. Apparently there is no style check rule for comments. 

> Is it safe to approve the commit anyway? I don't think it'll get pushed automatically...
Yep it is safe. Submitting a PR is a separate step, that is done after approval. I will do so in a bit.

**Pull request merged**